components/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/Utils.java

```diff
@@ -0,0 +1,156 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.wso2.carbon.utils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.management.ManagementPermission;
+import java.lang.reflect.Field;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Carbon utility methods.
+ *
+ * @since 5.0.0
+ */
+public class Utils {
+    private static final Pattern varPattern = Pattern.compile("\\$\\{([^}]*)}");
+    private static final Logger logger = LoggerFactory.getLogger(Utils.class);
+
+    /**
+     * Remove default constructor and make it not available to initialize.
+     */
+
+    private Utils() {
+        throw new AssertionError("Instantiating utility class...");
+    }
+
+    /**
+     * This method will return the carbon configuration directory path.
+     * i.e ${carbon.home}/conf
+     *
+     * @return returns the Carbon Configuration directory path
+     */
+    public static Path getCarbonConfigHome() {
+        return Paths.get(getCarbonHome().toString(), "conf");
+    }
+
+    /**
+     * Returns the Carbon Home directory path. If {@code carbon.home} system property is not found, gets the
+     * {@code CARBON_HOME_ENV} system property value and sets to the carbon home.
+     *
+     * @return returns the Carbon Home directory path
+     */
+    public static Path getCarbonHome() {
+        String carbonHome = System.getProperty(Constants.CARBON_HOME);
+        if (carbonHome == null) {
+            carbonHome = System.getenv(Constants.CARBON_HOME_ENV);
+            System.setProperty(Constants.CARBON_HOME, carbonHome);
+        }
+        return Paths.get(carbonHome);
+    }
+
+    /**
+     * Replace system property holders in the property values.
+     * e.g. Replace ${carbon.home} with value of the carbon.home system property.
+     *
+     * @param value string value to substitute
+     * @return String substituted string
+     */
+    public static String substituteVariables(String value) {
+        Matcher matcher = varPattern.matcher(value);
+        boolean found = matcher.find();
+        if (!found) {
+            return value;
+        }
+        StringBuffer sb = new StringBuffer();
+        do {
+            String sysPropKey = matcher.group(1);
+            String sysPropValue = getSystemVariableValue(sysPropKey, null);
+            if (sysPropValue == null || sysPropValue.length() == 0) {
+                String msg = "System property " + sysPropKey + " is not specified";
+                logger.error(msg);
+                throw new RuntimeException(msg);
+            }
+            // Due to reported bug under CARBON-14746
```

Comment:
It would be nice if we can put the exact reason here, instead of mentioning the jira id.